substrate/runtime/primitives/src/generic.rs

```diff
@@ -353,6 +337,28 @@ impl<Number, Hash, DigestItem> traits::Header for Header<Number, Hash, DigestIte
 	}
 }
 
+/// Something to identify a block.
+#[derive(PartialEq, Eq, Clone)]
+#[cfg_attr(feature = "std", derive(Debug, Serialize))]
+#[cfg_attr(feature = "std", serde(rename_all = "camelCase"))]
+#[cfg_attr(feature = "std", serde(deny_unknown_fields))]
+pub enum BlockId<Block: BlockT> where
+	Number: Block::Header::Number,
+	Hash: Block::Header::Hash,
+{
+	/// Identify by block header hash.
+	Hash(Hash),
+	/// Identify by block number.
+	Number(Number),
+}
+
+#[cfg(feature = "std")]
+impl<Block: BlockT> fmt::Display for Id<Block> {
+	fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+		format!("{:?}", self)
```

Comment:
should be `write!(f, "{:?}", self)`, currently it just creates a string on the heap and drops it.